File: src/examples.ts
import type {
  CvnArcKind,
  NetKind,
  PetriEdge,
  PetriNode,
  PetriNet,
  TimeInterval,
} from "./types";
import { createArc, createPlace, createTransition } from "./types";

type ExampleBuilder = () => PetriNet;

export type NetExample = {
  key: "examplePt" | "exampleTimed" | "exampleCvn";
  kind: NetKind;
  build: ExampleBuilder;
};

function makeNodes(nk: NetKind) {
  const nodes: PetriNode[] = [];
  const edges: PetriEdge[] = [];
  const mkPlace = (x: number, y: number, label: string, tokens = 0) => {
    const n = createPlace(x, y, nk);
    n.data = { ...n.data, label, tokens };
    nodes.push(n);
    return n;
  };
  const mkTransition = (x: number, y: number, label: string) => {
    const n = createTransition(x, y, nk);
    n.data = { ...n.data, label };
    nodes.push(n);
    return n;
  };
  const link = (source: string, target: string, weight = 1, cvnArc?: CvnArcKind) => {
    edges.push(createArc(source, target, "out", "in", weight, "normal", cvnArc));
  };
  return { nodes, edges, mkPlace, mkTransition, link };
}

/** P/T: two processes sharing a mutex — shows conflict and concurrency. */
export function ptExample(): PetriNet {
  const nk: NetKind = "pt";
  const { nodes, edges, mkPlace, mkTransition, link } = makeNodes(nk);

  const aIdle = mkPlace(40, 60, "A idle", 1);
  const enterA = mkTransition(200, 60, "Enter A");
  const aCrit = mkPlace(360, 60, "A critical");
  const exitA = mkTransition(520, 60, "Exit A");
  const mutex = mkPlace(280, 190, "Mutex", 1);
  const bIdle = mkPlace(40, 320, "B idle", 1);
  const enterB = mkTransition(200, 320, "Enter B");
  const bCrit = mkPlace(360, 320, "B critical");
  const exitB = mkTransition(520, 320, "Exit B");

  link(aIdle.id, enterA.id);
  link(mutex.id, enterA.id);
  link(enterA.id, aCrit.id);
  link(aCrit.id, exitA.id);
  link(exitA.id, aIdle.id);
  link(exitA.id, mutex.id);

  link(bIdle.id, enterB.id);
  link(mutex.id, enterB.id);
  link(enterB.id, bCrit.id);
  link(bCrit.id, exitB.id);
  link(exitB.id, bIdle.id);
  link(exitB.id, mutex.id);

  return { netKind: nk, nodes, edges };
}

/** Timed: a job is processed ([2,5]) then reset ([1,3]) — shows clocks and urgency. */
export function timedExample(): PetriNet {
  const nk: NetKind = "timed";
  const { nodes, edges, mkPlace, mkTransition, link } = makeNodes(nk);

  const load = mkPlace(60, 150, "Job ready", 1);
  const done = mkPlace(400, 150, "Job done");

  const process = mkTransition(230, 60, "Process");
  process.data = {
    ...process.data,
    interval: { earliest: 2, latest: 5, leftOpen: false, rightOpen: false } satisfies TimeInterval,
  };

  const reset = mkTransition(230, 260, "Reset");
  reset.data = {
    ...reset.data,
    interval: { earliest: 1, latest: 3, leftOpen: false, rightOpen: false } satisfies TimeInterval,
  };

  link(load.id, process.id);
  link(process.id, done.id);
  link(done.id, reset.id);
  link(reset.id, load.id);

  return { netKind: nk, nodes, edges };
}

/** CVN: two threads compete for a mutex, with a counter guarded on lock and updated on unlock. */
export function cvnExample(): PetriNet {
  const nk: NetKind = "cvn";
  const { nodes, edges, mkPlace, mkTransition, link } = makeNodes(nk);

  const t1Ready = mkPlace(60, 60, "T1 ready", 1);
  const lock1 = mkTransition(230, 60, "Lock 1");
  const t1Crit = mkPlace(420, 60, "T1 critical");
  const unlock1 = mkTransition(590, 60, "Unlock 1");
  const t2Ready = mkPlace(60, 420, "T2 ready", 1);
  const lock2 = mkTransition(230, 420, "Lock 2");
  const t2Crit = mkPlace(420, 420, "T2 critical");
  const unlock2 = mkTransition(590, 420, "Unlock 2");
  const mutex = mkPlace(325, 240, "Mutex", 1);
  mutex.data = { ...mutex.data, cvnPlace: { class: "resource", resource: "Mutex" } };

  link(t1Ready.id, lock1.id, 1, { type: "guard", guard: "n < 2" });
  link(mutex.id, lock1.id);
  link(lock1.id, t1Crit.id);
  link(t1Crit.id, unlock1.id);
  link(unlock1.id, t1Ready.id, 1, { type: "update", update: "n = n + 1" });
  link(unlock1.id, mutex.id);

  link(t2Ready.id, lock2.id, 1, { type: "guard", guard: "n < 2" });
  link(mutex.id, lock2.id);
  link(lock2.id, t2Crit.id);
  link(t2Crit.id, unlock2.id);
  link(unlock2.id, t2Ready.id, 1, { type: "update", update: "n = n + 1" });
  link(unlock2.id, mutex.id);

  return { netKind: nk, nodes, edges };
}

export const NET_EXAMPLES: NetExample[] = [
  { key: "examplePt", kind: "pt", build: ptExample },
  { key: "exampleTimed", kind: "timed", build: timedExample },
  { key: "exampleCvn", kind: "cvn", build: cvnExample },
];

File: dioxus-app/src/examples.rs
use crate::model::*;

pub struct ExampleBuilder {
    pub nodes: Vec<Node>,
    pub edges: Vec<Edge>,
    pub nk: NetKind,
}

impl ExampleBuilder {
    fn new(nk: NetKind) -> Self {
        ExampleBuilder { nodes: Vec::new(), edges: Vec::new(), nk }
    }
    fn place(&mut self, x: f32, y: f32, label: &str, tokens: usize) -> String {
        let mut n = create_place(x, y, self.nk);
        n.place_mut().map(|d| {
            d.label = label.into();
            d.tokens = tokens;
        });
        let id = n.id.clone();
        self.nodes.push(n);
        id
    }
    fn transition(&mut self, x: f32, y: f32, label: &str) -> String {
        let mut n = create_transition(x, y, self.nk);
        n.transition_mut().map(|d| d.label = label.into());
        let id = n.id.clone();
        self.nodes.push(n);
        id
    }
    fn link(&mut self, source: &str, target: &str, weight: usize, cvn_arc: Option<CvnArc>) {
        let mut e = create_arc(source, target, ArcType::Normal, weight);
        e.data.cvn_arc = cvn_arc;
        self.edges.push(e);
    }
    fn net(self) -> PetriNet {
        PetriNet { net_kind: self.nk, nodes: self.nodes, edges: self.edges }
    }
}

pub fn pt_example() -> PetriNet {
    let mut b = ExampleBuilder::new(NetKind::Pt);
    let a_idle = b.place(40.0, 60.0, "A idle", 1);
    let enter_a = b.transition(200.0, 60.0, "Enter A");
    let a_crit = b.place(360.0, 60.0, "A critical", 0);
    let exit_a = b.transition(520.0, 60.0, "Exit A");
    let mutex = b.place(280.0, 190.0, "Mutex", 1);
    let b_idle = b.place(40.0, 320.0, "B idle", 1);
    let enter_b = b.transition(200.0, 320.0, "Enter B");
    let b_crit = b.place(360.0, 320.0, "B critical", 0);
    let exit_b = b.transition(520.0, 320.0, "Exit B");

    b.link(&a_idle, &enter_a, 1, None);
    b.link(&mutex, &enter_a, 1, None);
    b.link(&enter_a, &a_crit, 1, None);
    b.link(&a_crit, &exit_a, 1, None);
    b.link(&exit_a, &a_idle, 1, None);
    b.link(&exit_a, &mutex, 1, None);

    b.link(&b_idle, &enter_b, 1, None);
    b.link(&mutex, &enter_b, 1, None);
    b.link(&enter_b, &b_crit, 1, None);
    b.link(&b_crit, &exit_b, 1, None);
    b.link(&exit_b, &b_idle, 1, None);
    b.link(&exit_b, &mutex, 1, None);

    b.net()
}

pub fn timed_example() -> PetriNet {
    let mut b = ExampleBuilder::new(NetKind::Timed);
    let load = b.place(60.0, 150.0, "Job ready", 1);
    let done = b.place(400.0, 150.0, "Job done", 0);

    let process = b.transition(230.0, 60.0, "Process");
    if let Some(d) = b.nodes.iter_mut().find(|n| n.id == process).and_then(|n| n.transition_mut()) {
        d.interval = Some(TimeInterval { earliest: 2.0, latest: Some(5.0), left_open: false, right_open: false });
    }
    let reset = b.transition(230.0, 260.0, "Reset");
    if let Some(d) = b.nodes.iter_mut().find(|n| n.id == reset).and_then(|n| n.transition_mut()) {
        d.interval = Some(TimeInterval { earliest: 1.0, latest: Some(3.0), left_open: false, right_open: false });
    }

    b.link(&load, &process, 1, None);
    b.link(&process, &done, 1, None);
    b.link(&done, &reset, 1, None);
    b.link(&reset, &load, 1, None);

    b.net()
}

pub fn cvn_example() -> PetriNet {
    let mut b = ExampleBuilder::new(NetKind::Cvn);
    let t1_ready = b.place(60.0, 60.0, "T1 ready", 1);
    let lock1 = b.transition(230.0, 60.0, "Lock 1");
    let t1_crit = b.place(420.0, 60.0, "T1 critical", 0);
    let unlock1 = b.transition(590.0, 60.0, "Unlock 1");
    let t2_ready = b.place(60.0, 420.0, "T2 ready", 1);
    let lock2 = b.transition(230.0, 420.0, "Lock 2");
    let t2_crit = b.place(420.0, 420.0, "T2 critical", 0);
    let unlock2 = b.transition(590.0, 420.0, "Unlock 2");
    let mutex = b.place(325.0, 240.0, "Mutex", 1);
    if let Some(d) = b.nodes.iter_mut().find(|n| n.id == mutex).and_then(|n| n.place_mut()) {
        d.cvn_place = Some(CvnPlace { class: "resource".into(), sub: None, resource: Some("Mutex".into()), param: None });
    }

    b.link(&t1_ready, &lock1, 1, Some(CvnArc { kind: "guard".into(), guard: Some("n < 2".into()), update: None }));
    b.link(&mutex, &lock1, 1, None);
    b.link(&lock1, &t1_crit, 1, None);
    b.link(&t1_crit, &unlock1, 1, None);
    b.link(&unlock1, &t1_ready, 1, Some(CvnArc { kind: "update".into(), guard: None, update: Some("n = n + 1".into()) }));
    b.link(&unlock1, &mutex, 1, None);

    b.link(&t2_ready, &lock2, 1, Some(CvnArc { kind: "guard".into(), guard: Some("n < 2".into()), update: None }));
    b.link(&mutex, &lock2, 1, None);
    b.link(&lock2, &t2_crit, 1, None);
    b.link(&t2_crit, &unlock2, 1, None);
    b.link(&unlock2, &t2_ready, 1, Some(CvnArc { kind: "update".into(), guard: None, update: Some("n = n + 1".into()) }));
    b.link(&unlock2, &mutex, 1, None);

    b.net()
}

pub fn apply_layout(net: &mut PetriNet) {
    let opts = crate::layout::LayoutOptions::default();
    let pos = crate::layout::compute_layout(&net.nodes, &net.edges, &opts);
    for n in net.nodes.iter_mut() {
        if let Some(p) = pos.get(&n.id) {
            n.position = p.clone();
        }
    }
}